jni_gen/java/src/main/java/com/github/dart_lang/jni_gen/apisummarizer/doclet/SummarizerDocletBase.java

```diff
@@ -0,0 +1,168 @@
+import com.github.dart_lang.jni_gen.apisummarizer.elements.Package;
+import com.github.dart_lang.jni_gen.apisummarizer.util.Log;
+import com.github.dart_lang.jni_gen.apisummarizer.util.SkipException;
+import java.util.*;
+import javax.lang.model.SourceVersion;
+import javax.lang.model.element.*;
+import javax.lang.model.util.ElementScanner9;
+import jdk.javadoc.doclet.Doclet;
+import jdk.javadoc.doclet.DocletEnvironment;
+import jdk.javadoc.doclet.Reporter;
+
+public class SummarizerDocletBase implements Doclet {
+  private AstEnv utils;
+
+  @Override
+  public void init(Locale locale, Reporter reporter) {}
+
+  @Override
+  public String getName() {
+    return "ApiSummarizer";
+  }
+
+  @Override
+  public Set<? extends Option> getSupportedOptions() {
+    return Collections.emptySet();
+  }
+
+  @Override
+  public SourceVersion getSupportedSourceVersion() {
+    return SourceVersion.RELEASE_11;
+  }
+
+  public static List<ClassDecl> types;
+
+  @Override
+  public boolean run(DocletEnvironment docletEnvironment) {
+    Log.timed("Initializing doclet");
+    utils = AstEnv.fromEnvironment(docletEnvironment);
+    SummarizingScanner p = new SummarizingScanner();
+    docletEnvironment.getSpecifiedElements().forEach(e -> p.scan(e, new SummaryCollector()));
+    types = p.types;
+    return true;
+  }
+
+  public static class SummaryCollector {
+    Stack<Package> packages = new Stack<>();
+    Stack<ClassDecl> types = new Stack<>();
+    Method method;
+  }
+
+  public class SummarizingScanner extends ElementScanner9<Void, SummaryCollector> {
+    List<Package> packages = new ArrayList<>();
+    List<ClassDecl> types = new ArrayList<>();
+    ElementBuilders builders = new ElementBuilders(utils);
+
+    // Each element in collector is a stack
+    // which is used to get topmost element
+    // and append the child to it.
+    // Eg: A variable element is always appended to topmost
+    // class
+    @Override
+    public Void scan(Element e, SummaryCollector collector) {
+      return super.scan(e, collector);
+    }
+
+    @Override
+    public Void visitPackage(PackageElement e, SummaryCollector collector) {
+      Log.verbose("Visiting package: %s", e.getQualifiedName());
+      collector.packages.push(new Package());
+      System.out.println("package: " + e.getQualifiedName());
+      var result = super.visitPackage(e, collector);
+      var collectedPackage = collector.packages.pop();
+      packages.add(collectedPackage);
+      return result;
+    }
+
+    @Override
+    public Void visitType(TypeElement e, SummaryCollector collector) {
+      // Supposedly it should visit every type only once.
+      // super.visitType visits the enclosed types and it should do that once.
+      // but its visiting nested classes twice.
+      // This could be a bug in my code, or something related to traversal order.
+      //
+      // Btw, the stack is useless now
+      // TODO: remove stack
```

Comment:
Please clean up all TODOs. Either (1) remove the TODO if it is not relevant, or (2) Make a GitHub issue with some background details.